src/mapgen_carpathian.cpp

```diff
@@ -0,0 +1,451 @@
+	float ster = NoisePerlin2D(&noise_step_terrain->np, x, z, seed);
+	float n_hills = NoisePerlin2D(&noise_hills->np, x, z, seed);
+	float n_ridge_mnt = NoisePerlin2D(&noise_ridge_mnt->np, x, z, seed);
+	float n_step_mnt = NoisePerlin2D(&noise_step_mnt->np, x, z, seed);
+
+	int height = -MAX_MAP_GENERATION_LIMIT;
+
+	for (s16 y = 1; y <= 30; y++) {
+		float mnt_var = NoisePerlin3D(&noise_mnt_var->np, x, y, z, seed);
+
+		// Gradient
+		int grad = (1 - y);
+		// Shallow seabed
+		if (y < 1) {
+			grad *= 3;
+		}
+
+		// Hill/Mountain height (hilliness)
+		float hill1 = getLerp(height1, height2, mnt_var);
+		float hill2 = getLerp(height3, height4, mnt_var);
+		float hill3 = getLerp(height3, height2, mnt_var);
+		float hill4 = getLerp(height1, height4, mnt_var);
+		float hilliness = fmax(fmin(hill1, hill2), fmin(hill3, hill4));
+
+		// Rolling hills
+		float hill_mnt = hilliness * pow(n_hills, 2.f);
+		float hills = pow(hter, 3.f) * hill_mnt;
+
+		// Ridged mountains
+		float ridge_mnt = hilliness * (1.f - fabs(n_ridge_mnt));
+		float ridged_mountains = pow(rter, 3.f) * ridge_mnt;
+
+		// Step (terraced) mountains
+		float step_mnt = hilliness * getSteps(n_step_mnt, fabs(mnt_var));
+		float step_mountains = pow(ster, 3.f) * step_mnt;
+
+		// Final terrain level
+		float mountains = hills + ridged_mountains + step_mountains;
+		float density = ground + mountains + grad;
+
+		if (y > density && height < 0)
+			height = y;
+	}
+
+	return height;
+}
+
+
+///////////////////////////////////////////////////////////////////////////////
+
+
+int MapgenCarpathian::generateTerrain()
+{
+	MapNode mn_air(CONTENT_AIR);
+	MapNode mn_stone(c_stone);
+	MapNode mn_water(c_water_source);
+
+	s16 stone_surface_max_y = -MAX_MAP_GENERATION_LIMIT;
+	u32 index2d = 0;
+	u32 index3d = 0;
+
+	// Calculate noise for terrain generation
+	noise_base->perlinMap2D(node_min.X, node_min.Z);
+	noise_height1->perlinMap2D(node_min.X, node_min.Z);
+	noise_height2->perlinMap2D(node_min.X, node_min.Z);
+	noise_height3->perlinMap2D(node_min.X, node_min.Z);
+	noise_height4->perlinMap2D(node_min.X, node_min.Z);
+	noise_hills_terrain->perlinMap2D(node_min.X, node_min.Z);
+	noise_ridge_terrain->perlinMap2D(node_min.X, node_min.Z);
+	noise_step_terrain->perlinMap2D(node_min.X, node_min.Z);
+	noise_hills->perlinMap2D(node_min.X, node_min.Z);
+	noise_ridge_mnt->perlinMap2D(node_min.X, node_min.Z);
+	noise_step_mnt->perlinMap2D(node_min.X, node_min.Z);
+	noise_mnt_var->perlinMap3D(node_min.X, node_min.Y - 1, node_min.Z);
+
+	//// Place nodes
+	for (s16 z = node_min.Z; z <= node_max.Z; z++) {
+		for (s16 y = node_min.Y - 1; y <= node_max.Y + 1; y++) {
+			u32 vi = vm->m_area.index(node_min.X, y, z);
+			for (s16 x = node_min.X; x <= node_max.X;
+					x++, vi++, index2d++, index3d++) {
+				if (vm->m_data[vi].getContent() != CONTENT_IGNORE)
+					continue;
+
+				// Base terrain
+				float ground = noise_base->result[index2d];
+
+				// Gradient
+				int grad = (1 - y);
```

Comment:
Seems here you are using y = 1 (sea level) as the grad zero point and compressing the terrain variation below sea level.
Sea level is actually variable using the global mapgen parameter `water_level` so:
```
+				int grad = (water_level - y);
// Shallow seabed
+				if (y < water_level) {
+					grad *= 3;
+				}
```
This must be why it was reported that changing water_level messes stuff up.